Deep Dive rewrite: align every page to the current architecture (#43)
FhenixProtocol/fhenix-developer-docs: deep-dive/cofhe-components/overview.mdx

        end
    end

    subgraph Registry["Registry chain"]
        CR["CommitmentRegistry<br/>op commitments"]
    end

    FHEC --> TM
    SDK -- "encrypt" --> ZK
    SDK -- "decrypt" --> TEE
    SDK -- "publish result" --> TM
    TM --> PS
    TM --> L
    TEE -- "ACL check" --> TM
    TEE --> CTS
    TEE -- "verify commitment" --> CR
    ZK --> CTS
    ENG <--> CTS
    P --> CR
    PRT --> TEE
    PRT --> ZK

    style Yours stroke-dasharray: 6 4
    style PRT stroke-dasharray: 6 4
```

## The onchain contracts

- **FHE.sol**: the Solidity library your contract imports. It exposes arithmetic, comparison, and logical operations on encrypted values, plus access control (`allow`, `allowGlobal`) and decrypt-result verification.
- **[TaskManager](/deep-dive/cofhe-components/task-manager)**: the gateway for all FHE operation requests. It validates requests, emits task events for the offchain services, and manages permissions through the ACL.
- **[ACL](/deep-dive/cofhe-components/acl)**: the access control contract. It records who may use or decrypt each handle; all writes to it go through the TaskManager.

> - **[PlaintextsStorage](/deep-dive/cofhe-components/plaintext-storage)**: stores decrypted results after they are published onchain with a verified Teecryptor signature.

Reviewer: ```suggestion
- **[PlaintextsStorage](/deep-dive/cofhe-components/plaintext-storage)**: stores decrypted results; TaskManager writes them during `publishDecryptResult`, only after ECDSA-verifying the result against `decryptResultSigner`."
```

> - **[CommitmentRegistry](/deep-dive/cofhe-components/commitment-registry)**: lives on a dedicated registry chain and records a hash commitment for every ciphertext the coprocessor produces or verifies. Teecryptor checks it before decrypting anything.

Reviewer: ```suggestion
- **[CommitmentRegistry](/deep-dive/cofhe-components/commitment-registry)**: lives on a dedicated registry chain and records a hash commitment for every ciphertext the coprocessor produces or verifies. Teecryptor verifies against that before decrypting anything.
```

Author: Will do

## The offchain components

- **[Client SDK](/client-sdk/introduction/overview)** (`@cofhe/sdk`): the TypeScript library applications use to encrypt inputs, manage [permits](/client-sdk/guides/permits), and decrypt outputs.

> - **[ZK Verifier](/deep-dive/cofhe-components/zk-verifier)**: a standalone service that verifies encrypted inputs in batches. It checks each zero-knowledge proof, stores the ciphertexts, and signs one approval for the whole batch.

Reviewer: no mention about TEE here? only down? cuz in Teecryptor there is a mention

- **[FHE Engine](/deep-dive/cofhe-components/fhe-engine)**: the execution pipeline. It subscribes to TaskManager events, validates and orders operations, runs them on encrypted data, and posts a commitment for every result.

> - **Ciphertext store**: the database that holds every ciphertext, fronted by the CT Server. The ZK Verifier writes verified inputs, the FHE Engine reads operands and writes results, and Teecryptor fetches bytes to decrypt.

Reviewer: do we really wanna go that deep into ct-server and etc?

Author: Mmm, why not? I

---
title: "CoFHE Architecture Overview"
description: "How CoFHE's onchain contracts and offchain services fit together to run FHE computations for any EVM chain"
---

CoFHE (Coprocessor for Fully Homomorphic Encryption) lets smart contracts compute on encrypted data. Contracts request operations through onchain calls; a set of offchain services executes them and commits to the results. Data stays encrypted until an authorized decryption, and every result can be verified against an onchain commitment.

```mermaid

> %%{init: {"theme": "base", "themeVariables": {"fontFamily": "Menlo, Monaco, Consolas, monospace", "fontSize": "16px", "primaryColor": "#8FBAF5", "primaryBorderColor": "#2E7CF6", "primaryTextColor": "#0A1626", "lineColor": "#4C8DFF", "signalColor": "#4C8DFF", "signalTextColor": "#8FA3BF", "actorBkg": "#8FBAF5", "actorBorder": "#2E7CF6", "actorTextColor": "#0A1626", "actorLineColor": "#3D4654", "noteBkgColor": "#14171C", "noteBorderColor": "#3D4654", "noteTextColor": "#AFC3DE", "activationBkgColor": "#1E3A5F", "activationBorderColor": "#4C8DFF", "clusterBkg": "#14171C", "clusterBorder": "#3D4654", "titleColor": "#E7EAEE", "edgeLabelBackground": "#8FBAF5", "textColor": "#AFC3DE", "labelTextColor": "#E7EAEE", "tertiaryColor": "#14171C", "loopTextColor": "#AFC3DE", "labelBoxBkgColor": "#1E3A5F", "labelBoxBorderColor": "#4C8DFF"}, "sequence": {"actorFontFamily": "Menlo, Monaco, Consolas, monospace", "messageFontFamily": "Menlo, Monaco, Consolas, monospace", "noteFontFamily": "Menlo, Monaco, Consolas, monospace", "width": 220, "actorFontSize": 16, "messageFontSize": 16, "noteFontSize": 15}}}%%

Reviewer 2: <img width="120" height="150" alt="Image" src="https://github.com/user-attachments/assets/d1aba45d-1710-4a0b-a4df-c75fe72c36ca" />

<img width="2720" height="3400" alt="Image" src="https://github.com/user-attachments/assets/4a42875d-d995-4676-ada5-764ad295b785" />

Reviewer 2: This is my suggested structure

Author: But what about that we want to strive and have everything as a public repo and code, we might want to expose service by service rather than packing everything under FHE Engine. I agree that this is much more clearer then explaining each service - it just depends on what we want to have in the docs I think this is something worth input from @YonatanE22 and @liorbond.

Author: The diagram now matches your structure, including "Your contract" in its own dashed group outside CoFHE. The open question is still service-by-service versus one FHE Engine block, which needs Yonatan's and Lior's input.

Author: Update, and it walks back part of my last reply. I rendered both versions side by side. Yours is cleaner, and the gap is not something I can tune away: your layout is hand-placed, Mermaid uses automatic layout. In our render the edge labels detached and floated mid-canvas, the nested TEE enclave boxes I had just added made it worse, and one label was clipped by the ciphertext store cylinder.

So I simplified instead of copying. The three color-coded flows are gone, since they needed a legend to read, and six redundant edge labels are gone with them. The TEE detail moved into the node text rather than a nested box. Five edge labels are left: encrypt, decrypt, publish result, ACL check, verify commitment, which is close to your set.

It is cleaner than what I pushed this morning, and it is still not as clean as yours. We are accepting that rather than shipping a designed SVG for this page.

    linkStyle 7,8 stroke:#7FDB8F
    linkStyle 3,4,5,6,12,13 stroke:#0AD9DC
    linkStyle 14,15 stroke:#E0C36A
```

## The onchain contracts

- **FHE.sol**: the Solidity library your contract imports. It exposes arithmetic, comparison, and logical operations on encrypted values, plus access control (`allow`, `allowGlobal`) and decrypt-result verification.
- **[TaskManager](/deep-dive/cofhe-components/task-manager)**: the gateway for all FHE operation requests. It validates requests, emits task events for the offchain services, and manages permissions through the ACL.
- **[ACL](/deep-dive/cofhe-components/acl)**: the access control contract. It records who may use or decrypt each handle; all writes to it go through the TaskManager.
- **[PlaintextsStorage](/deep-dive/cofhe-components/plaintext-storage)**: stores decrypted results after they are published onchain with a verified Teecryptor signature.
- **[CommitmentRegistry](/deep-dive/cofhe-components/commitment-registry)**: lives on a dedicated registry chain and records a hash commitment for every ciphertext the coprocessor produces or verifies. Teecryptor checks it before decrypting anything.

## The offchain services

> - **[Client SDK](/client-sdk/introduction/overview)** (`@cofhe/sdk`): the TypeScript library applications use to encrypt inputs, manage [permits](/client-sdk/guides/permits), and decrypt outputs.

Reviewer 2: I don't think all the services, even with a deep dive, are relevant to users.
When we give the users the option to run cofhe by themselves, it will be relevant, but not now.
We do need to keep the things they interact with, or big blocks for the overview.
The list IMO needs to match the diagram suggestion in the comment above.

Author: I wrote a response on the diagram suggestion. Let's focus the discussion there on the scope of the overview and the deep-dive docs. If we drop the services from the overview, how do we elaborate on them in the other sections? That would mean removing the other pages as well, right?

Reviewer 2: Right

Author: Done. The overview now lists big blocks only, and the per-service pages for slim-listener and fheos-server are gone, folded into one FHE Engine page with redirects.

> - **[Client SDK](/client-sdk/introduction/overview)** (`@cofhe/sdk`): the TypeScript library applications use to encrypt inputs, manage [permits](/client-sdk/guides/permits), and decrypt outputs.

Reviewer 2: The client sdk isn't a service

Author: Agree on the terminology

> - **[ZK Verifier](/deep-dive/cofhe-components/zk-verifier)**: verifies the zero-knowledge proof attached to every encrypted input, signs it, and stores the ciphertext bytes.

Reviewer 2: The zk verifier is a service now, I think that it's still in the time when it wasn't

Author: I will rewrite the title as mentioned. I want to reflect the current state of the system.

- **[Slim Listener](/deep-dive/cofhe-components/slim-listener)**: watches TaskManager events on each host chain and forwards them into the coprocessor's queues.
- **[FheOS Server](/deep-dive/cofhe-components/fheos-server)**: ingests task events, verifies inputs, creates result placeholders, and routes work to the FHE Engine. It does not execute FHE operations itself.
- **[FHE Engine](/deep-dive/cofhe-components/fhe-engine)**: executes every FHE operation on encrypted data, persists the resulting ciphertexts, and emits a commitment for each result.
- **CT Server**: the gate to the ciphertext database. Every ciphertext read and write for the other services goes through it.
- **Blockchain Poster**: batches result commitments and posts them to the CommitmentRegistry.
- **[Teecryptor](/deep-dive/cofhe-components/teecryptor)**: the decryption service. It runs inside a hardware-attested TEE, authorizes every request against the onchain ACL, verifies the ciphertext commitment, and returns signed or sealed results.

> - **Partners**: independent custodians of the key material. Each holds a Shamir share and releases it only to an attested enclave. See [Key Management](/deep-dive/cofhe-components/key-management).

Reviewer 2: Partners don't run a service

Author: Right, I should rewrite the title to "off-chain components" rather than "services."

Services communicate through message queues rather than direct calls, so each stage can retry and scale independently.

## Data flows

1. **[Encryption request](/deep-dive/data-flows/encryption-request-flow)**: an input is encrypted client-side and proven valid with a zero-knowledge proof before it enters the chain.

> 2. **[FHE operation flow](/deep-dive/data-flows/fhe-operation-request-flow)**: a contract requests a computation; the coprocessor executes it and commits to the result.

Reviewer 2: No need the "flow" word, it's already in "data flows" headline

Author: 100%

3. **[Decryption](/deep-dive/data-flows/decryption-request-flow)**: the SDK asks Teecryptor to decrypt a handle, either as a signed plaintext for onchain publication or sealed to a permit for offchain reads.

## Future plans

> Decryption is currently performed by Teecryptor inside a hardware-attested TEE. A multi-party [Threshold Network](/deep-dive/research/future-plans) is the planned successor.

Reviewer 2: The zk verification is also part of the plan when converting to TN

Author: Cool, didn't know that. Do you have any more info we can write here about that?

Reviewer 2: The partners will run zk verification on their side, but I don't know yet what it will look like in MPC.
The idea is to move the two TEEs to MPC at the same time.

Author: Written up. future-plans.mdx now has a ZK Verifier row in the trust-point table, and the section is retitled "The path to multi-party computation": both TEE services move together, and each partner runs ZK verification on its own side. I also dropped the line that called the ZK TEE a removed training wheel, since the enclave is still a trust point.
